contrib/check-whitespace.sh

```diff
@@ -24,8 +24,7 @@ file_patterns='
 *Makefile
 '
 
-# TODO: Look also for trailing empty lines, and missing '\n' after the last line
-if git --no-pager grep --color -n --full-name -e ' $' -- $file_patterns; then
+if git --no-pager grep --color -n --full-name -P '[\s ]+$' -- $file_patterns; then
```

Comment:
That should do the same thing, right?
```suggestion
if git --no-pager grep --color -n --full-name -P '\s+$' -- $file_patterns; then
```

Reply:
It should but it doesn't—git does not consider nbsp to be a space. In general git's regex matching here is very broken. In basic or POSIX mode, it treats file contents as Latin-1, which is wrong (took me a while to debug that). In Perl-compatible mode it at least treats the text as UTF-8, so that works but it doesn't correctly. However, even in Perl-compatible mode `\s` doesn't match nbsp. So this uses the literal nbsp character in a class. Of course there are other space characters that still aren't matched, which isn't great but those seem pretty unlikely to sneak into the code base.

Reply:
I'm sorry I asked! 😂

Reply:
Of course, I could have done something more drastic like rewrite this code to use grep or perl, but that seemed harder and worse than just doing this nasty hack.

Reply:
This might be my fault. When you run code via a Makefile, we force it into `C` locale (aka latin-1), whereas if you run it directly, you are probably using `en_US.UTF-8` or something similar.

https://github.com/JuliaLang/julia/blob/03433a20a848ee1c07cfcae6283ec25e01b49a18/Make.inc#L100-L102

Reply:
Ah yes, the classic James Mickens situation: "I have no tools because I have destroyed my tools with my tools."

Reply:
Ah, well, I didn't see this until now and I've rewritten the check script in Julia, which is easier to maintain and extend if we want to check for other things. Good to know that this isn't due to git being completely flaky, but rather due to UNIX and POSIX having a completely awful locale system, which was a bad idea, but at least we all know it's a bad idea.